File: breadith-first-search/bfs.js
/**
 * Binary Search Tree
 */
const tree = {
	10: {
		value: 10,
		left: 4,
		right: 17,
	},
	4: {
		value: 4,
		left: 1,
		right: 9,
	},
	17: {
		value: 17,
		left: 12,
		right: 18,
	},
	1: {
		value: 1,
		left: null,
		right: null,
	},
	9: {
		value: 9,
		left: null,
		right: null,
	},
	12: {
		value: 12,
		left: null,
		right: null,
	},
	18: {
		value: 18,
		left: null,
		right: null,
	},
};

/**
 * Breadith-First-Search - Iterative bfs approach
 * @param {object | map | hashtable} tree - Illustrative Binary Search Tree varying by language
 * @param {object | map | hashtable} root - Binary search tree root/entry point
 * @param {integer} target - The desired value
 * O(V+E) where V is number of vertices in the graph and E is number of edges in the graph.
 */
function bfs(tree, root, target) {
	const queue = [];

	// add root node to the queue
	queue.push(root);

	while (queue.length > 0) {
		const currentNode = queue[0];
		// return if target node is found
		if (currentNode.value === target) {
			return target;
		}

		// if current node contains children
		// push to the queue
		if (currentNode.left !== null) {
			queue.push(tree[currentNode.left]);
		}

		// if current node contains children
		// push to the queue
		if (currentNode.right !== null) {
			queue.push(tree[currentNode.right]);
		}

		// remove current node from the queue
		queue.shift();
	}

	// not found
	return -1;
}

// prints 12
console.log(bfs(tree, tree[10], 12));

// prints -1
console.log(bfs(tree, tree[10], 120));

File: breadith-first-search/bfs.go
package main

import "fmt"

// Root of the tree
type Root struct {
	root *Node
}

// tree node aka vertex
type Node struct {
	left  *Node
	value int
	right *Node
}

/**
 * Binary Search Tree
 */
var tree Root = Root{
	root: &Node{
		value: 10,
		left: &Node{
			value: 4,
			left: &Node{
				value: 1,
				left:  nil,
				right: nil,
			},
			right: &Node{
				value: 9,
				left:  nil,
				right: nil,
			},
		},
		right: &Node{
			value: 17,
			left: &Node{
				value: 12,
				left:  nil,
				right: nil,
			},
			right: &Node{
				value: 18,
				left:  nil,
				right: nil,
			},
		},
	},
}

// Remove element of the array without mutating the position of the other elements
func removeIndex(s []*Node, index int) []*Node {
	return append(s[:index], s[index+1:]...)
}

/**
 * Breadith-First-Search - Iterative bfs approach
 * @param {object | map | hashtable} tree - Illustrative Binary Search Tree varying by language
 * @param {object | map | hashtable} root - Binary search tree root/entry point
 * @param {integer} target - The desired value
 */
func bfs(root *Node, target int) int {
	if root == nil {
		return -1
	}

	var queue []*Node
	queue = append(queue, root)

	for len(queue) > 0 {
		currentNode := queue[0]

		if currentNode.value == target {
			return target
		}

		if currentNode.left != nil {
			queue = append(queue, currentNode.left)
		}

		if currentNode.right != nil {
			queue = append(queue, currentNode.right)
		}

		// Essentially a shift/pop of the front of the queue/array
		queue = removeIndex(queue, 0)
	}

	return -1
}

func main() {
	// prints 12
	fmt.Println(bfs(tree.root, 12))

	// prints -1
	fmt.Println(bfs(tree.root, 120))

	emptyTree := Root{
		root: nil,
	}
	// prints -1
	fmt.Println(bfs(emptyTree.root, 12))
}
